src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,290 @@
+
+        /// <summary>
+        /// Initialize any environmentally derived state after specialization, prior to starting the publisher.
+        /// </summary>
+        internal void Initialize()
+        {
+            _metricPublishInterval = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.MetricsPublishIntervalMS);
+            _initialPublishDelay = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.InitialPublishDelayMS);
+            _intervalStopwatch = ValueStopwatch.StartNew();
+            MetricsFilePath = _optionsMonitor.CurrentValue.MetricsFilePath;
+
+            IsAlwaysReady = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsAlwaysReadyInstance) == "1";
+        }
+
+        internal async Task OnPublishMetrics()
+        {
+            if (FunctionExecutionCount == 0 && FunctionExecutionTimeMS == 0 && !IsAlwaysReady)
+            {
+                // no activity to report
+                return;
+            }
+
+            // we've been accumulating function activity for the entire period
+            // publish this activity and reset
+            Metrics metrics = null;
+            lock (_lock)
+            {
+                metrics = new Metrics
+                {
+                    TotalTimeMS = (long)_intervalStopwatch.GetElapsedTime().TotalMilliseconds,
+                    ExecutionCount = FunctionExecutionCount,
+                    ExecutionTimeMS = FunctionExecutionTimeMS,
+                    IsAlwaysReady = IsAlwaysReady
+                };
+
+                FunctionExecutionTimeMS = FunctionExecutionCount = 0;
+                _instanceActivityStopwatch = ValueStopwatch.StartNew();
+            }
+
+            await PublishMetricsAsync(metrics);
+        }
+
+        private async void OnFunctionMetricsPublishTimer(object state)
+        {
+            await OnPublishMetrics();
+        }
+
+        private async Task PublishMetricsAsync(Metrics metrics)
+        {
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile(_fileSystem))
+                {
+                    return;
+                }
+
+                string metricsContent = JsonConvert.SerializeObject(metrics);
+                _logger.PublishingMetrics(metricsContent);
+
+                if (metricsPublishEnabled)
+                {
+                    string fileName = $"{Guid.NewGuid().ToString().ToLower()}.json";
+                    string filePath = Path.Combine(MetricsFilePath, fileName);
+
+                    using (var streamWriter = _fileSystem.File.CreateText(filePath))
+                    {
+                        await streamWriter.WriteAsync(metricsContent);
+                    }
+                }
+            }
+            catch (Exception ex) when (!ex.IsFatal())
+            {
+                _logger.LogError(ex, "Error writing metrics file.");
```

Comment:
The `CreateDirectory` API could throw for a [few reasons](https://learn.microsoft.com/en-us/dotnet/api/system.io.directory.createdirectory?view=net-6.0#system-io-directory-createdirectory(system-string)). Should we also log the parameter passed to it as well (`MetricsFilePath`) ?